### PR TITLE
Zero copy for send

### PR DESCRIPTION
**Pull request by ice1000**
A subtask of #222 , extracted.

### CODE REVIEW

File: src/call/mod.rs [hunk 1]
```diff
@@ -225,12 +387,15 @@ impl Call {
     ) -> Result<BatchFuture> {
         let _cq_ref = self.cq.borrow()?;
         let i = if initial_meta { 1 } else { 0 };
-        let f = check_run(BatchType::Finish, |ctx, tag| unsafe {
+        let f = check_run_with_content(BatchType::Finish, msg.to_vec(), |content,
```
[the remark this answers is not in the document]
Comment (Hoverbear) [reply src/call/mod.rs]: Maybe you can add a new function, we can mark the old as deprecated and give some time for users to upgrade.

[the remark this answers is not in the document]
Comment (ice1000) [reply src/call/mod.rs]: @Hoverbear This function is invisible to user, and when you look at the call-side, you'll know that there will always be a copy (the call-side is reusing the buffer).

**Comment by siddontang** [comment]
Great work !!!

**Comment by ice1000** [comment]
Tests still pass when this path is applied:

```diff
diff --git a/src/call/mod.rs b/src/call/mod.rs
index dcb7530..4aa1c9b 100644
--- a/src/call/mod.rs
+++ b/src/call/mod.rs
@@ -670,11 +670,19 @@ impl StreamingBase {
 }

 /// Flags for write operations.
-#[derive(Default, Clone, Copy)]
+#[derive(Clone, Copy)]
 pub struct WriteFlags {
     flags: u32,
 }

+impl Default for WriteFlags {
+    fn default() -> Self {
+        let ret = WriteFlags { flags: 0 };
+        ret.buffer_hint(true);
+        ret
+    }
+}
+
```

**Comment by ice1000** [comment]
Is there any reason that line in this code

https://github.com/ice1000/grpc-rs/blob/c4f1c665bdeeca6eb4b7c42ba39a68ad9dbc858a/tests/cases/misc.rs#L48-L53

can fail? Why does nightly build succeed? What if we do a `thread::sleep` here?

**Comment by ice1000** [comment]
@overvenus Asked me to explain why there's no use-after-free when `buffer hint` flag is enabled.

**Causion: C++**

First let's use CLion's "find usages" function to find all the "read access" of `grpc_op.flags`:

image

We can easily find that only these two usages need to be reviewed:
image

## Part 1

For the first one (at line 1689), variable `flags` is passed in expression

```c
call->sending_stream.Init(
     &op->data.send_message.send_message->data.raw.slice_buffer, flags)
```

We go to the definition of `Init`, and we can easily understand the implementation -- it's calling a placement new operator which forwards `flags` to the second parameter of the constructor of `call->sending_stream`:

```cpp
public:
  template <typename... Ts>
  void Init(Ts&&... args) {
    new (&space_) Type(std::forward<Ts>(args)...);
  }
private:
  typename std::aligned_storage<sizeof(Type), alignof(Type)>::type space_;
```

Then we can easily find the constructor being called in `Init`:

```cpp
class SliceBufferByteStream : public ByteStream {
 public:
  // Removes all slices in slice_buffer, leaving it empty.
  SliceBufferByteStream(grpc_slice_buffer* slice_buffer, uint32_t flags);
```

`flags` is passed to its superclass' constructor and assigned to `flags_` field:

```cpp
 protected:
  ByteStream(uint32_t length, uint32_t flags)
      : length_(length), flags_(flags) {}

 private:
  const uint32_t length_;
  uint32_t flags_;
```

`flags_` is exposed as readonly by method `flags()`:

```cpp
  uint32_t flags() const { return flags_; }
```

This method has 11 usages:

image

When we ignore those trivial copies and buffer-hint-unrelated usages (they're mostly about compressing), we have only two usages left:

image

This fact (it's used in these two places) is confirmed via looking for the usage of `GRPC_WRITE_BUFFER_HINT` defined in `grpc_types.h`.

In the first usage, it's initiating writing when buffer hint is not set:

```cpp
if (!(op->send_message &&
      (op->payload->send_message.send_message->flags() &
       GRPC_WRITE_BUFFER_HINT))) {
  grpc_chttp2_initiate_write(
      t, GRPC_CHTTP2_INITIATE_WRITE_SEND_INITIAL_METADATA);
}
```

In the second usage, it's used to check whether we're buffering or not when we've finished passing the message:

```c
if (flags & GRPC_WRITE_BUFFER_HINT) {
  s->next_message_end_offset -= t->write_buffer_size;
  s->write_buffering = true;
} else {
  s->write_buffering = false;
}
continue_fetching_send_locked(t, s);
maybe_become_writable_due_to_send_msg(t, s);
```

We can see that the value of buffer hint is moved to `s->write_buffering`, which is only used in `maybe_become_writable_due_to_send_msg`:

```cpp
static void maybe_become_writable_due_to_send_msg(grpc_chttp2_transport* t,
                                                  grpc_chttp2_stream* s) {
  if (s->id != 0 && (!s->write_buffering ||
                     s->flow_controlled_buffer.length > t->write_buffer_size)) {
    grpc_chttp2_mark_stream_writable(t, s);
    grpc_chttp2_initiate_write(t, GRPC_CHTTP2_INITIATE_WRITE_SEND_MESSAGE);
  }
}
```

We can see that it's sending buffer's content when buffer hint is not enabled or buffered message is larger than `write_buffer_size`. But none of these are about actually sending the buffered messages.

**Comment by ice1000** [comment]
## Part 2

These are all related to metadata, so we'll still need to look at the previous part

**Comment by ice1000** [comment]
I'll start another PR later since maintaining one PR derived from another PR is too painful, closing.

**Pull request closed without merging**